### ISSUE TITLE
Typed GDScript doesn't like Resource keys in Dictionary

### ISSUE THREAD

**Issue by razcore-art** [godotengine/godot#25088]
**Godot version:** v3.1.beta1.official

**OS/device including version:** Ubuntu 18.10

**Issue description:**
```gdscript
var test : = preload("res://Test.tres")

func _ready() -> void:
	var a : = { test: 0 } # no error
	print(a) # no error
	print(a[test]) # error: Invalid index type (Resource) for base 'Dictionary'
```

If we have instead:
```gdscript
var test = preload("res://Test.tres") # note this is dynamic now

func _ready() -> void:
	var a : = { test: 0 } # no error
	print(a) # no error
	print(a[test]) # no error
```

**Comment by razcore-art** [comment]
This might or might not be related, but this also happens:

```gdscript
func _ready() -> void:
	var s : = {{"t": 5}: 7} # no error
	var a : = {"t": 5}
	s[a] = 7 # no error
	s[a] # error if s[a] = [...] isn't used first
	s[{"t": 5}] # error
```

I'm not even sure if `Dictionary` is serializable/hash-able to be able to be used as keys. I know in `Python` you can't use them as keys.

**Comment by vnen** [comment]
This is a duplicate of #23520, should be already fixed in 3.1-beta2.

The second post is a runtime error, it's expected since a dictionary's hash is based on the reference, not on the contents.
